bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,521 @@
+        try {
+            return stream.read();
+        } catch (IOException e) {
+            return -1;
+        }
+    }
+
+    private void receiveInBinaryProtocol(SerialPortEvent serialPortEvent) {
+        int b = 0;
+
+        // Read the start byte
+        if (!inMessage) // skip until 0x7E
+        {
+            b = 0;
+            while (b != 0x7E && b != -1) {
+                b = readByte(in);
+            }
+            if (b == -1) {
+                return;
+            }
+
+            inMessage = true;
+            messageBufferLength = 0;
+        }
+        logger.trace("CaddxCommunicator.handleBinaryProtocol() Got start byte");
+
+        // Read the message length
+        if (messageBufferLength == 0) {
+            b = readByte(in);
+            if (b == -1) {
+                return;
+            }
+
+            messageBufferLength = b + 2; // add two bytes for the checksum
+            message = new byte[messageBufferLength];
+        }
+        logger.trace("CaddxCommunicator.handleBinaryProtocol() Got message length {}", b);
+
+        // Read the message
+        while (true) {
+            b = readByte(in);
+            if (b == -1) {
+                return;
+            }
+            message[messageBufferIndex] = (byte) b;
+
+            if (message[messageBufferIndex] == 0x7D) {
+                unStuff = true;
+                continue;
+            }
+
+            if (unStuff) {
+                message[messageBufferIndex] |= 0x20;
+                unStuff = false;
+            }
+
+            messageBufferIndex++;
+            if (messageBufferIndex == messageBufferLength) {
+                break;
+            }
+        }
+        logger.trace("CaddxCommunicator.handleBinaryProtocol() Got message {}", message[0]);
+
+        // Received data
+        CaddxMessage caddxMessage = new CaddxMessage(message, true);
+
+        try {
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() Exchanging");
+
+            if (!exchanger.offer(caddxMessage, 3, TimeUnit.SECONDS)) {
+                logger.debug("Offered message was not received");
+            }
+        } catch (InterruptedException e) {
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() InterruptedException caught.");
+            Thread.currentThread().interrupt();
+        }
+
+        // Initialize state for next reception
+        inMessage = false;
+        haveMessageLength = false;
+        haveFirstByte = false;
+        messageBufferLength = 0;
+        message = new byte[0];
+        messageBufferIndex = 0;
+        unStuff = false;
+    }
+
+    private int readAsciiByte(InputStream stream) {
+        if (!haveFirstByte) { // this is the 1st digit
+            int b = readByte(in);
+            if (b == -1) {
+                return -1;
+            }
+            tempAsciiByte = (b - 0x30) * 0x10;
+            haveFirstByte = true;
+        }
+
+        if (haveFirstByte) { // this is the 2nd digit
```

Comment:
Could you explain why you need this boolean field? From what I can tell (with the exception of an EOF), the readAsciiByte method is always called with haveFirstByte=false and will always exit leaving haveFirstByte=false.
Is this flag specifically meant to handle recovering from an EOF?

Reply:
Yes. It is for the EOF handling.
Two bytes of input constitute an output byte. e.g. output byte 0x0A is transmitted as "0x30 0x41" [Ascii 0 (48), Ascii A (65)].